### PR TITLE
[EWL-6262] Subcat Stacking Bug

### PR DESCRIPTION
**Pull request by JaeBea**
## Ticket(s)
**Jira Ticket**
- [EWL-6262 | Subcat Stacking Bug](https://issues.ama-assn.org/browse/EWL-6262)

## Description
Corrects the spacing or individual subcategory items in the subcategory exploration component so that they are evenly spaced and left aligned. Cleans up logic for determining number of subcat items per row.

## To Test
- Pull `bugfix/EWL-6262-subcat-stacking-bug` branch
- Run `gulp serve` in ama-style-guide-2/styleguide to ensure a local version of style guide is running.
- This work affects visuals in D8 and should be tested against it to ensure there are no regressions. Update ama-d8 branch and local provision vars accordingly. 
- Go to any category page with more than one row of subcat items, such as [Delivering Care](http://ama-one.local/delivering-care) and confirm that when clicking view more, the subcategory items show left aligned when less than max possible items are on a line.
- Test this in IE 11, Safari, and Firefox to ensure that with updates everything works same as before.

## Visual Regressions
A report is available [here](http://htmlpreview.github.io/?https://github.com/AmericanMedicalAssociation/ama-style-guide-2/blob/visual-regression-testing-artifact/bugfix/EWL-6262-subcat-stacking-bug/html_report/index.html).

## Relevant Screenshots/GIFs
Screenshot from Drupal

## Remaining Tasks
N/A

## Additional Notes
N/A

### CODE REVIEW

**Comment by nstriedinger** [comment]
The View Less doesn't seem to work.

screen shot 2018-11-13 at 12 15 43 pm

File: styleguide/source/assets/js/subcategory.js [hunk 1]
```diff
@@ -12,54 +12,69 @@
...
-          subcategory.slice(0, 2).css('display', 'block');
-        } else if ((subcategoryWrapper > 700 && subcategoryWrapper < 1000) && subcategoryTitle < 200) {
-          subcategory.slice(0, 3).css('display', 'block');
-        } else if ((subcategoryWrapper > 1000 && subcategoryWrapper < 1200) && subcategoryTitle < 200) {
-          subcategory.slice(0, 4).css('display', 'block');
-        } else {
-          subcategory.slice(0, 5).css('display', 'block');
+        // Set subcategory row items to lowest that should display.
+        var subcategoryItemsPerRow = Math.floor((subcategoryExplorationWidth - subcategoryTitle)/subcategoryItemWidth);
+        if(subcategoryItemsPerRow < 2) {
```
Comment (nstriedinger) [inline styleguide/source/assets/js/subcategory.js]:
```suggestion
        if (subcategoryItemsPerRow < 2) {
```

File: styleguide/source/assets/js/subcategory.js [hunk 2]
```diff
@@ -12,54 +12,69 @@
...
+          subcategoryItemsPerRow = 2;
         }
+
+        // Get the computed styles and set all items margins to the browser computed ones
+        var computedSubcatStyles = window.getComputedStyle($subcategory[0], null);
+
+        $subcategory.css({
+          'margin-left': computedSubcatStyles.marginLeft,
+          'margin-right': computedSubcatStyles.marginRight,
+
```
Comment (nstriedinger) [inline styleguide/source/assets/js/subcategory.js]:
```suggestion
```

File: styleguide/source/assets/js/subcategory.js [hunk 3]
```diff
@@ -12,54 +12,69 @@
...
+          $viewMore.hide();
+          $subcategoryContainer.addClass('expanded');
+          $viewLess.show();
+
         });
 
         $('.viewLess').click(function(e) {
           e.preventDefault();
-          subcategory.hide();
+          $element.hide();
```
Comment (nstriedinger) [inline styleguide/source/assets/js/subcategory.js]: Element undefined. Maybe?
```suggestion
          $viewLess.hide();
```

**Comment by nstriedinger** [comment]
A little borked in IE11

screen shot 2018-11-13 at 12 28 43 pm

File: styleguide/source/assets/js/subcategory.js (hunk 3 above)
Comment (JaeBea) [reply styleguide/source/assets/js/subcategory.js]: Yes, the $element var is not defined. I needed to change this to $subcategory.

**Comment by JaeBea** [comment]
@nstriedinger ,
I have updated this to manually calculate the margins and use set widths for non-mobile views. This insures IE11 behaves consistently.

**Comment by nstriedinger** [comment]
mmm -- still looks a little borked in IE11

screen shot 2018-11-14 at 9 34 17 am

**Comment by JaeBea** [comment]
@nstriedinger ,
I have added a minimal amount of spacing for items. This should resolve the issue in IE.

**Comment by nstriedinger** [comment]
Is this ok?
screen shot 2018-11-14 at 11 53 52 am

**Review by damontgomery — CHANGES_REQUESTED** [review]
I'm seeing this go to 2 rows on page load before showing all.

This only happens in Drupal, not with the style guide which works as expected.

two-rows

http://ama-one.local/advocacy

**Comment by JaeBea** [comment]
@damontgomery ,
I have made a change that seems to fix the issues in D8.

@nstriedinger ,
I tried to recreate the issue you mentioned after updating the subcat width to a static number and was not able to reproduce the issue in IE. Please try this again to confirm if my change has resolved the issue.

**Comment by damontgomery** [comment]
@JaeBea, have there been updates since my comment? I still see the issue before showing all the items in D8.

**Comment by JaeBea** [comment]
@damontgomery ,
Apologies,  the commit did not get pushed. Please try again.

**Comment by nstriedinger** [comment]
This is about 1100 x 850 in IE11. Is this right?

screen shot 2018-11-15 at 11 50 44 am

**Review by damontgomery — DISMISSED** [review]
This works as described.

I tested with the style guide page http://localhost:3000/?p=pages-category and the Drupal page http://ama-one.local/advocacy. I tested on Chrome and on IE11.

The testing steps were very helpful.

**Review by Brian-Clement — CHANGES_REQUESTED** [review]
@JaeBea I'm seeing a regression on the `bp-small` breakpoint:

image

Also the `bp-med` viewport is showing 2 now instead of 3:

image

This is appearing both in my Drupal theme as well as in Pattern Lab.

I tested this by merging SG2 `develop` into this branch, running `gulp serve` and testing both PL and Drupal.

**Comment by JaeBea** [comment]
@Brian-Clement , @damontgomery , and @nstriedinger ,
The method used to handle this has changed.  This should hopefully prevent any issues in IE11 and maintain proper spacing at tablet viewport widths that are affected by rounding errors. Crossing fingers.

**Review by damontgomery — APPROVED** [review]
Tested working in Chrome (MacOS), Safari, IE 11.

Tested with http://ama-one.local/advocacy

I'm noticing a few issues that are not for this feature, but surrounding it. I don't think these need to be addressed here, but I thought it might be helpful to bring them up so we can mention we want to ignore them.

On Chrome tablet simulator (iPad) using the d8 site, there is no right padding

tablet_padding

Using Chrome phone simulator, there is no spacing on the top.

phone-top-spacing

**Pull request merged**